std/zip.d

```diff
@@ -1,61 +1,98 @@
- * Example:
- * ---
-// Read existing zip file.
-import std.digest.crc, std.file, std.stdio, std.zip;
-
-void main(string[] args)
+Read and write data in the
+$(LINK2 https://en.wikipedia.org/wiki/Zip_%28file_format%29, zip archive)
+format.
+
+Standards:
+
+The current implementation mostly conforms to
+$(LINK2 https://www.iso.org/standard/60101.html, ISO/IEC 21320-1:2015),
+which means,
+$(UL
+$(LI that files can only be stored uncompressed or using the deflate mechanism,)
+$(LI that encryption features are not used,)
+$(LI that digital signature features are not used,)
+$(LI that patched data features are not used, and)
+$(LI that archives may not span multiple volumes.)
+)
+
+Additionally, archives are checked for malware attacks and rejected if detected.
+This includes
+$(UL
+$(LI $(LINK2 https://news.ycombinator.com/item?id=20352439, zip bombs) which
+     generate gigantic amounts of unpacked data)
+$(LI zip archives that contain overlapping records)
+$(LI chameleon zip archives which generate different unpacked data, depending
+     on the implementation of the unpack algorithm)
+)
+
+The current implementation makes use of the zlib compression library.
+
+Usage:
+
+There are two main ways of usage: Extracting files from a zip archive
+and storing files into a zip archive. These can be mixed though (e.g.
+read an archive, remove some files, add others and write the new
+archive).
+
+Examples:
+
+Example for reading an existing zip archive:
+---
+import std.stdio : writeln, writefln;
+import std.file : read;
+import std.zip;
+
+// read a zip file into memory
+auto zip = new ZipArchive(read(args[1]));
+
+// iterate over all zip members
+writefln("%-10s  %-8s  Name", "Length", "CRC-32");
+foreach (name, am; zip.directory)
 {
-    // read a zip file into memory
-    auto zip = new ZipArchive(read(args[1]));
-    writeln("Archive: ", args[1]);
-    writefln("%-10s  %-8s  Name", "Length", "CRC-32");
-    // iterate over all zip members
-    foreach (name, am; zip.directory)
-    {
-        // print some data about each member
-        writefln("%10s  %08x  %s", am.expandedSize, am.crc32, name);
-        assert(am.expandedData.length == 0);
-        // decompress the archive member
-        zip.expand(am);
-        assert(am.expandedData.length == am.expandedSize);
-    }
+    // print some data about each member
+    writefln("%10s  %08x  %s", am.expandedSize, am.crc32, name);
+    assert(am.expandedData.length == 0);
+
+    // decompress the archive member
+    zip.expand(am);
+    assert(am.expandedData.length == am.expandedSize);
 }
+---
 
-// Create and write new zip file.
+Example for writing files into a zip archive:
+---
 import std.file : write;
 import std.string : representation;
+import std.zip;
+
+// Create an ArchiveMembers for each file.
+ArchiveMember file1 = new ArchiveMember();
+file1.name = "test1.txt";
+file1.expandedData("Test data.\n".dup.representation);
+file2.compressionMethod = CompressionMethod.none; // don't compress
```

Comment:
Need to change `file2` to `file1` to make this runnable. Maybe the example should be made into a documented unit test, although adding cleanup code to delete the generated zip file would make the example slightly worse. (Worse because it includes code that does nothing to make the demonstration clearer but exists solely for the convenience of the unit test framework.)

Reply:
Corrected the bug. Thanks for noting.